torch/csrc/LazyTensor.cpp

```diff
@@ -0,0 +1,300 @@
+  std::vector<const LazyTensorImpl*> inputs;
+  while (frontier.size()) {
+    std::vector<c10::IValue> new_frontier;
+    for (const auto& ival : frontier) {
+      if (ival.isTensor()) {
+        auto t = ival.toTensor();
+        if (t.is_lazy()) {
+          auto* lt = static_cast<LazyTensorImpl*>(t.unsafeGetTensorImpl());
+          if (lt->schema_) {
+            // make into Item, push to record
+            record.emplace_back(
+                std::make_pair(lt->schema_->name(), lt->index_));
+            // maybe add to new_frontier
+            new_frontier.reserve(new_frontier.size() + lt->inps_.size());
+            new_frontier.insert(
+                new_frontier.end(), lt->inps_.begin(), lt->inps_.end());
+            continue;
+          } else {
+            inputs.emplace_back(lt);
+          }
+        }
+      }
+      record.emplace_back(ival);
+    }
+    frontier = new_frontier;
+  }
+
+  auto ge = graph_cache.get(record);
+  if (ge) {
+    return std::make_pair(ge, inputs);
+  } else {
+    inputs.clear();
+  }
+
+  auto g = std::make_shared<torch::jit::Graph>();
+  torch::jit::Value* output;
+  std::tie(output, inputs) = getValue(lt_, g);
+  g->registerOutput(output);
+  ge = std::make_shared<torch::jit::GraphExecutor>(g);
+  graph_cache.put(std::move(record), ge);
+
+  return std::make_pair(ge, inputs);
+}
+
+Tensor LazyTensorImpl::to_eager() const {
+  if (!schema_) {
+    auto t = inps_[0].toTensor();
+    return t;
+  }
+
+  std::shared_ptr<torch::jit::GraphExecutor> ge;
+  std::vector<const LazyTensorImpl*> inputs;
+  std::tie(ge, inputs) = getGraph(this);
+  torch::jit::Stack stack;
+  for (auto& inp : inputs) {
+    stack.push_back(inp->inps_[0]);
+  }
+
+  ge->run(stack);
+
+  // We registered only the output we care about
+  auto output_t = stack[0].toTensor();
+  if (output_t.is_variable()) {
+    return output_t;
+  }
+  auto output_var = torch::autograd::make_variable(output_t);
+  output_t = output_var.tensor_data();
+  TORCH_INTERNAL_ASSERT(output_t.is_variable());
+  return output_t;
+}
+
+} // namespace at
+
+namespace at {
+
+using namespace torch::jit;
+void lazyRun(const OperatorHandle& op, torch::jit::Stack* s) {
+  const FunctionSchema& fs = op.schema();
+  if (fs.name() == "aten::to_eager") {
+    auto t = pop(*s).toTensor();
+    auto lt = static_cast<LazyTensorImpl*>(t.unsafeGetTensorImpl());
+    auto tensor = lt->to_eager();
+    push(*s, tensor);
+    return;
+  }
+  auto arg_len = fs.arguments().size();
+  auto ret_len = fs.returns().size();
+  auto args = last(*s, arg_len);
+  if (ret_len == 1) {
+    auto tensor = detail::make_tensor<LazyTensorImpl>(std::move(fs), args.vec());
+    auto var = torch::autograd::make_variable(tensor);
+    drop(*s, arg_len);
+    push(*s, var);
+  } else {
+    drop(*s, arg_len);
+    std::vector<IValue> ret_vec;
+    for (auto i = 0; i < ret_len; ++i) {
+      auto tensor = detail::make_tensor<LazyTensorImpl>(std::move(fs), args.vec(), i);
```

Comment:
This looks dangerous too... won't each output individually recompute the entire `fs(args)` operation? NB: I haven't actually read your graph processing code above, maybe you "fix it up" there. But even if you do, that's a pretty big code smell to me: rather than destroy sharing and have to reconstruct it later, LazyTensor should *preserve* sharing from the get go, so you don't have to work hard later.

Reply:
(the std::move is straight up broken, gotta fix that)

what do you mean by sharing?  If there are two outputs, it'll be recomputed for each, but that can easily change going forward

Reply:
Sharing = not recomputing. In another one of your responses, you mention that there's some sort of cache involved. But I don't want a cache! If I reuse a computation multiple times in my source program, with or without caching, I should reuse the computation in lazy mode.

Reply:
I'll ensure the computation is computed once

I'd like to highlight that the reason for laziness is for the user to give up some more control to optimization paths, which should be free to recompute if semantically equivalent.

Reply:
Sure. But we should faithfully record the original sharing/recomputation that the user wrote down, before we start busting it open :)